drv/stm32h7-update-server/src/main.rs

```diff
@@ -0,0 +1,342 @@
+enum Trace {
+    EraseStart,
+    EraseEnd,
+    WriteStart,
+    WriteEnd,
+    FinishStart,
+    FinishEnd,
+    WriteBlock(usize),
+    None,
+}
+
+enum UpdateState {
+    NoUpdate,
+    InProgress,
+    Finished,
+}
+
+ringbuf!(Trace, 64, Trace::None);
+
+const FLASH_TIMEOUT: usize = 10000;
+
+struct ServerImpl<'a> {
+    flash: &'a device::flash::RegisterBlock,
+    state: UpdateState,
+}
+
+impl<'a> ServerImpl<'a> {
+    // See RM0433 Rev 7 section 4.3.13
+    fn swap_banks(&mut self) -> Result<(), RequestError<UpdateError>> {
+        let flash = unsafe { &*device::FLASH::ptr() };
+
+        ringbuf_entry!(Trace::FinishStart);
+        if flash.optsr_cur().read().swap_bank_opt().bit() {
+            flash
+                .optsr_prg()
+                .modify(|_, w| w.swap_bank_opt().clear_bit());
+        } else {
+            flash.optsr_prg().modify(|_, w| w.swap_bank_opt().set_bit());
+        }
+
+        flash.optcr().modify(|_, w| w.optstart().set_bit());
+
+        loop {
+            if !flash.optsr_cur().read().opt_busy().bit() {
+                break;
+            }
+        }
+
+        ringbuf_entry!(Trace::FinishEnd);
+        Ok(())
+    }
+
+    fn poll_flash_done(&mut self) -> Result<(), RequestError<UpdateError>> {
+        let mut cnt = 0;
+        let mut seen = false;
+
+        loop {
+            if self.flash.bank2().sr.read().qw().bit() {
+                seen = true;
+                cnt = 0;
+            }
+
+            if seen && !self.flash.bank2().sr.read().qw().bit() {
+                break;
+            } else {
+                cnt += 1;
+            }
+
+            if cnt == FLASH_TIMEOUT {
+                break;
+            }
+        }
+
+        if cnt == FLASH_TIMEOUT {
+            return Err(UpdateError::Timeout.into());
+        }
+
+        if self.flash.bank2().sr.read().dbeccerr().bit() {
+            return Err(UpdateError::EccDoubleErr.into());
+        }
+
+        if self.flash.bank2().sr.read().sneccerr1().bit() {
+            return Err(UpdateError::EccSingleErr.into());
+        }
+
+        if self.flash.bank2().sr.read().rdserr().bit() {
+            return Err(UpdateError::SecureErr.into());
+        }
+
+        if self.flash.bank2().sr.read().rdperr().bit() {
+            return Err(UpdateError::ReadProtErr.into());
+        }
+
+        if self.flash.bank2().sr.read().operr().bit() {
+            return Err(UpdateError::WriteEraseErr.into());
+        }
+
+        if self.flash.bank2().sr.read().incerr().bit() {
+            return Err(UpdateError::InconsistencyErr.into());
+        }
+
+        if self.flash.bank2().sr.read().strberr().bit() {
+            return Err(UpdateError::StrobeErr.into());
+        }
+
+        if self.flash.bank2().sr.read().pgserr().bit() {
+            return Err(UpdateError::ProgSeqErr.into());
+        }
+
+        if self.flash.bank2().sr.read().wrperr().bit() {
+            return Err(UpdateError::WriteProtErr.into());
```

Comment:
I wonder if it's worth having an error type that can describe more than one of these cases. Like, potentially one that contains the literal SR value.)

Reply:
So I took a look at this and unfortunately I don't think we can get the full SR value. If we use `idol` we need the error value to be representable as `u16` The bits we care about in SR can only be represented as `u16` (too many for `u8`). I like the idea of doing this long term but for now I think the specific enums are better.